include/ginkgo/core/config/property_tree.hpp

```diff
@@ -0,0 +1,209 @@
+     *
+     * @param boolean  the bool type value
+     */
+    pnode(bool boolean);
+
+    /**
+     * Constructor for integer with all integer type
+     *
+     * @tparam T  input type
+     *
+     * @param integer  the integer type value
+     */
+    template <typename T,
+              std::enable_if_t<std::is_integral<T>::value>* = nullptr>
+    pnode(T integer);
+
+    /**
+     * Constructor for string
+     *
+     * @param str  string type value
+     */
+    pnode(const std::string& str);
+
+    /**
+     * Constructor for char* (otherwise, it will use bool)
+     *
+     * @param str  the string like "..."
+     */
+    pnode(const char* str);
+
+    /**
+     * Constructor for double
+     *
+     * @param real  the floating point type value
+     */
+    pnode(double real);
+
+    /**
+     * Constructor for array
+     *
+     * @param array  an pnode array
+     */
+    pnode(const array_type& array);
+
+    /**
+     * Constructor for map
+     *
+     * @param map  a (string, pnode)-map
+     */
+    pnode(const map_type& map);
+
+    /**
+     * bool conversion. It's true if and only if it is not empty.
+     */
+    operator bool() const noexcept;
+
+    /**
+     * Get the current node status.
+     *
+     * @return the status
+     */
+    status_t get_status() const;
+
+    /**
+     * Get the array. It will throw error if the current node does not hold an
+     * array.
+     *
+     * @return the array const reference
+     */
+    const array_type& get_array() const;
+
+    /**
+     * Get the map. It will throw error if the current node does not hold an
+     * map.
+     *
+     * @return the map const reference
+     */
+    const map_type& get_map() const;
+
+    /**
+     * Get the boolean value. It will throw error if the current node does not
+     * hold an boolean value.
+     *
+     * @return the boolean value
+     */
+    bool get_boolean() const;
+
+    /**
+     * Get the integer value. It will throw error if the current node does not
+     * hold an integer value.
+     *
+     * @return the integer value with type int64_t
+     */
+    std::int64_t get_integer() const;
+
+    /**
+     * Get the real floating point value. It will throw error if the current
+     * node does not hold an real value.
+     *
+     * @return the real value with type double
+     */
+    double get_real() const;
+
+    /**
+     * Get the string. It will throw error if the current node does not hold an
+     * string.
+     *
+     * @return the string
+     */
+    std::string get_string() const;
+
+    /**
+     * This function is to access the data under the map. It will throw error
+     * when it does not hold a map. When access non-existent path in the map, it
+     * will return an empty node.
+     *
+     * @param path  the key of the map
+     *
+     * @return node const reference. If the map does not have the path, return
+     * an empty node.
+     */
+    const pnode& get(const std::string& path) const;
```

Comment:
nit: maybe call the parameter `key`? I think path has a different meaning for json objects, so it might be a bit confusing.